File: maki-code-index/src/bash.rs
use tree_sitter::Node;

use crate::common::{LanguageExtractor, Section, SkeletonEntry, node_text, truncate};

pub(crate) struct BashExtractor;

impl BashExtractor {
    fn extract_function(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))?;
        Some(SkeletonEntry::new(
            Section::Function,
            node,
            format!("{name}()"),
        ))
    }

    fn extract_variable(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))?;
        if !name.chars().all(|c| c.is_ascii_uppercase() || c == '_') {
            return None;
        }
        let value = node
            .child_by_field_name("value")
            .map(|n| format!(" = {}", truncate(node_text(n, source), 60)))
            .unwrap_or_default();
        Some(SkeletonEntry::new(
            Section::Constant,
            node,
            format!("{name}{value}"),
        ))
    }
}

impl LanguageExtractor for BashExtractor {
    fn extract_nodes(&self, node: Node, source: &[u8], _attrs: &[Node]) -> Vec<SkeletonEntry> {
        match node.kind() {
            "function_definition" => self.extract_function(node, source).into_iter().collect(),
            "variable_assignment" => self.extract_variable(node, source).into_iter().collect(),
            _ => Vec::new(),
        }
    }

    fn is_doc_comment(&self, _node: Node, _source: &[u8]) -> bool {
        false
    }
}

File: plugins/index/lang/bash.lua
return function(U)
  local get_text = U.get_text
  local new_entry = U.new_entry
  local truncate = U.truncate
  local SECTION = U.SECTION

  return {
    import_separator = "/",

    extract_nodes = function(node, source, _attrs)
      local kind = node:type()

      if kind == "function_definition" then
        local name_node = node:field("name")[1]
        if not name_node then return {} end
        return { new_entry(SECTION.Function, node, get_text(name_node, source) .. "()") }

      elseif kind == "variable_assignment" then
        local name_node = node:field("name")[1]
        if not name_node then return {} end
        local name = get_text(name_node, source)
        if not name:match("^[A-Z_]+$") then return {} end
        local val_node = node:field("value")[1]
        local val_str = val_node and (" = " .. truncate(get_text(val_node, source), 60)) or ""
        return { new_entry(SECTION.Constant, node, name .. val_str) }
      end

      return {}
    end,
  }
end
